### PR TITLE
FOPTS-2607 Deployment of rightsize azure managed disks policy

### PR DESCRIPTION
**Pull request by midir99**
### Description

Deploy first version of Rightsize Azure Managed Disks policy.

### Issues Resolved

- https://flexera.atlassian.net/browse/FOPTS-2607

### Contribution Check List

- [ ] New functionality includes testing.
- [X] New functionality has been documented in the README if applicable
- [X] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/azure/rightsize_managed_disks/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Azure Rightsize Managed Disks
+
+## What it does
+
+This Policy Template scans all volumes in the given account and identifies any volume that meets the user-specified criteria for being oversized. The user can filter volumes based on usage percentage of capacity (GiB used), usage percentage of IOPS, usage percentage of throughput, or any combination of these. Any volumes that meet the user-specified criteria are considered oversized. If any oversized volumes are found, an incident report will show the volumes and related information. An email will be sent to the user-specified email addresses.
+
+- It is important that you keep constant LUN numbers for your data disks because when policy retrieves metrics for them, it will only look for the latest LUN number assigned to the data disk, if this one is changed multiple times before running the policy, only the data points using the current LUN the disk has will be retrieved for analysis.
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/README.md:7]: There are multiple sentences here separated by commas instead of periods or semicolons and a grammatical error. This is what it should look like:

It is important that you keep constant LUN numbers for your data disks because when policy retrieves metrics for them. It will only look for the latest LUN number assigned to the data disk; if changed multiple times before running the policy, only the data points using the current LUN the disk has will be retrieved for analysis.

File: cost/azure/rightsize_managed_disks/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- *Allow/Deny Subscriptions* - Determines whether the Allow/Deny Subscriptions List parameter functions as an allow list (only providing results for the listed subscriptions) or a deny list (providing results for all subscriptions except for the listed subscriptions).
+- *Allow/Deny Subscriptions List* - A list of allowed or denied Subscription IDs/names. If empty, no filtering will occur and recommendations will be produced for all subscriptions.
+- *Allow/Deny Regions* - Whether to treat Allow/Deny Regions List parameter as allow or deny list. Has no effect if Allow/Deny Regions List is left empty.
+- *Allow/Deny Regions List* - Filter results by region, either only allowing this list or denying it depending on how the above parameter is set. Leave blank to consider all the regions.
+- *Minimum used disk space percentage to consider a managed disk for downsizing* - The minimum used disk space percentage threshold at which to consider a disk to be 'oversized' and therefore be flagged for downsizing.
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/README.md:26]: Name of parameter should be shorter and capitalised. The description is for the additional details.  This feedback applies to the other parameters and the policy itself as well.

Example of how it should look:

- *Minimum Used Disk Space (%)* - The minimum used disk space percentage threshold at which to consider a disk to be 'oversized' and therefore be flagged for downsizing.

File: cost/azure/rightsize_managed_disks/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+- [**Azure Resource Manager Credential**](https://docs.flexera.com/flexera/EN/Automation/ProviderCredentials.htm#automationadmin_109256743_1124668) (*provider=azure_rm*) which has the following permissions:
+
+  - `Microsoft.Compute/disks/read`
+  - `Microsoft.Compute/virtualMachines/read`
+  - `Microsoft.Insights/metrics/read`
+
+- [**Flexera Credential**](https://docs.flexera.com/flexera/EN/Automation/ProviderCredentials.htm) (*provider=flexera*).
+
+This is only required for the application of the metaparent policy.
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/README.md:51]: This should be modified to match other policies. Also, we should use this credential to pull the policy name and include that in the title of the incident, similar to our other cost recommendation policies. See Azure Rightsizing Compute for an example:

```# Get applied policy metadata for use later
datasource "ds_applied_policy" do
  request do
    auth $auth_flexera
    host rs_governance_host
    path join(["/api/governance/projects/", rs_project_id, "/applied_policies/", policy_id])
    header "Api-Version", "1.0"
  end
end

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 4]
```diff
@@ -0,0 +1,1081 @@
...
+    },
+    headers: {
+      "User-Agent": "RS Policies",
+    },
+    ignore_status: [400, 403, 404],
+  }
+EOS
+end
+
+datasource "ds_azure_virtual_machines_with_disk_space" do
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:450]: Why is this here? Looks like it does nothing but return an empty list

File: cost/azure/rightsize_managed_disks/README.md (hunk 1 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/README.md:7]: Thank you @XOmniverse , I applied these changes!

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt (hunk 4 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:450]: Hey @XOmniverse, currently we want to leave this "empty datasource" that is just returning an empty list, this would be replaced by an API call to download a file containing a JSON list that would contain information about the disk remaining capacity (GiB), something like this:

image

So since this change will be applied soon (we just need to verify a couple things with client), we would like to leave that code in there since we already tested it and make sure it works.

I would add something to the README.md telling that this feature is not yet available.

File: cost/azure/rightsize_managed_disks/README.md (hunk 3 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/README.md:51]: Thank you @XOmniverse , I applied these changes!

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 5]
```diff
@@ -0,0 +1,1135 @@
...
+EOS
+end
+
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy "pol_azure_rightsize_managed_disks" do
+  validate $ds_disk_rightsizing_recommendations do
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:908]: We should do a validate_each instead and have the policy check on each instance in a list. See other recommendations policies, such as the Azure Unused Volumes policy, for information on how to do this.

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 6]
```diff
@@ -0,0 +1,1135 @@
...
+    summary_template "{{ data.policyName }}: There are {{ len data.recommendations }} oversized Azure Managed Disks, you could save ${{ data.monthlySavings }} USD!"
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:909]: This should match the formatting of other similar policies. For example, from the Azure Unused Volumes policy:

summary_template "{{ with index data 0 }}{{ .policy_name }}{{ end }}: {{ len data }} Azure Unused Volumes Found"

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 7]
```diff
@@ -0,0 +1,1135 @@
...
+    detail_template <<-'EOS'
+## Summary
+Found {{ len data.recommendations }} Azure Managed Disks that are *oversized* during the last {{ parameters.param_stats_lookback }} days. *Oversized* disks are those which have **less or equal than {{ parameters.param_min_used_disk_iops_pct }}% {{ parameters.param_stats_aggregation_for_disk_iops }} IOPS utilization, less or equal than {{ parameters.param_min_used_disk_throughput_pct }}% {{ parameters.param_stats_aggregation_for_disk_throughput }} throughput utilization and less or equal than {{ parameters.ds_min_used_disk_space_pct }}% used capacity (GiB)** within the last {{ parameters.param_stats_lookback }} days.
```
Comment (XOmniverse) [inline cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:912]: This should be formatted to look like the other recommendations policies and the message should largely be constructed in the final datasource. For example, from the Azure Unused Volumes policy:

    detail_template <<-'EOS'
    **Potential Monthly Savings:** {{ with index data 0 }}{{ .total_savings }}{{ end }}

    {{ with index data 0 }}{{ .message }}{{ end }}
    EOS

**Review by XOmniverse — CHANGES_REQUESTED** [review]

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 8]
```diff
@@ -0,0 +1,1135 @@
...
+      if (currentUsedIops <= maxIops && currentUsedThroughput <= maxThroughput) {
+        diskIopsAndThroughputDontFitInDiskSize = false
+      }
+      suggestedDiskSize += 1
+    }
+    return {
+      pricePerUnit: Math.round(
+          (premSsdV2Pricing.PREMIUM_LRS_PROVISIONED_CAPACITY.pricePerUnit * 24 * 30 * suggestedDiskSize)
+        + (premSsdV2Pricing.PREMIUM_LRS_PROVISIONED_IOPS.pricePerUnit * 24 * 30 * currentUsedIops)
+        + (premSsdV2Pricing.PREMIUM_LRS_PROVISIONED_THROUGHPUT_MBPS.pricePerUnit * 24 * 30 * currentUsedThroughput)
```
Comment (Rangelmv23) [inline cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:727]: If this multiplication is to get the hours of the month, multiply the value by 730 instead of by 24 and then by 30.

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt (hunk 5 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt]: Thank you for this comment!

I addressed this, now it's using validate each as other rightsize policies :)

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt (hunk 6 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt]: Thank you for this comment!

I addressed this, now it's using the same format as other rightsize policies :)

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt (hunk 7 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt]: Thanks for this comment!

I addressed this by constructing most of the message in the previous `datasource`, in this case I'm using strictly USD for the currency since all the APIs this policy is working with only use dollars. I also tried to simplify the logic of the message construction and improve the readability for the user putting in separate points what were the parameters he specified.

Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt]: image
image

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt (hunk 8 above)
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt]: Thank you @Rangelmv23 ! I addressed this comment :)

**Review by XOmniverse — CHANGES_REQUESTED** [review]
Just some small changes

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 9]
```diff
@@ -0,0 +1,1160 @@
...
+      field "throughputPctAverage" do
+        label "Throughput Average %"
+      end
+      field "throughputThreshold" do
+        label "Throughput Threshold %"
+      end
+      field "sizeGB" do
+        label "Actual Disk Size GiB"
+      end
+      # Hide until we launch a version of the policy where the clients provide a datasource
```
[the remark this answers is not in the document]
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:1041]: @XOmniverse, thank you, I removed it.

File: cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt [hunk 10]
```diff
@@ -0,0 +1,1160 @@
...
+      service: disk.service.split("/")[0],
+    })
+
+    totalMonthlySavings += monthlySavings
+  }
+
+  if (totalMonthlySavings >= param_min_savings) {
+    monthlySavings = monthlySavings.toFixed(2)
+
+    recommendations[0].message = "Oversized disks are those that meet the following conditions within the last " + param_stats_lookback + " day(s):\n\n"
```
[the remark this answers is not in the document]
Comment (midir99) [reply cost/azure/rightsize_managed_disks/azure_rightsize_managed_disks.pt:894]: @XOmniverse, thank you, I added that condition in that IF statement as the first one to be evaluated.

**Review by night-codes — APPROVED** [review]

**Review by XOmniverse — APPROVED** [review]
